syntax = "proto3";

package tp_c2s;

import "shared.proto";

message ConnectedToRoom {
	message Client {
		string uuid = 1;
		string name = 2;
	}

	repeated Client clients = 1;
}

message CreateGameRoom {
	string name = 1;
}

message PlayerSlot {
	enum SlotType {
		UNSPECIFIED_SLOT_TYPE = 0;
		HUMAN = 1;
		AI = 2;
		OPEN_SLOT = 4;
		CLOSED_SLOT = 5;
	}

	SlotType slot_type = 1;
	string name = 2;
	int32 index = 3;
}

message StartGame {
	bool ready = 1;
}

message JoinGameRoom {
	string server_uuid = 1;
}

message GameCommand {
	bool pause = 1;
}

message BoardMove {
	string player_uuid = 1;
	tp.Move move = 2;
}

message BoardExternalSquares {
	string uuid = 1;
	repeated tp.BlockType block_types = 2;
}

message BoardNextBlock {
	string uuid = 1;
	tp.BlockType next = 2;
}

message GameRestart {
	tp.BlockType current = 1;
	tp.BlockType next = 2;
}

message FailedToConnect {
}

message FromServer {

}

message Wrapper {
	PlayerSlot player_slot = 1;
	StartGame start_game = 2;
	ConnectedToRoom connected_to_room = 3;
	GameCommand game_command = 4;
	BoardMove board_move = 5;
	BoardExternalSquares board_external_squares = 6;
	BoardNextBlock next_block = 7;
	GameRestart game_restart = 8;
	JoinGameRoom Join_game_room = 9;
	FailedToConnect failed_to_connect = 10;
	CreateGameRoom create_game_room = 11;
}
